References
==========

**Bibliography**

Drost, H.G., Gabel, A., Liu, J., Quint, M. and Grosse, I. (2018). **myTAI: evolutionary transcriptomics with R.**
*Bioinformatics*, **34(9)**. [https://doi.org/10.1093/bioinformatics/btx835](https://doi.org/10.1093/bioinformatics/btx835)

Emms, D.M. and Kelly, S. (2019). **OrthoFinder: phylogenetic orthology inference for comparative genomics.** 
*Genome biology*, **20(1)**. [https://doi.org/10.1186/s13059-019-1832-y](https://doi.org/10.1186/s13059-019-1832-y)

Huerta-Cepas, J., Serra, F. and Bork, P. (2016). **ETE 3: reconstruction, analysis, and visualization of phylogenomic data.**
*Molecular biology and evolution*, **33(6)**. [https://doi.org/10.1093/molbev/msw046](https://doi.org/10.1093/molbev/msw046)

Packer, J.S., Zhu, Q., Huynh, C., Sivaramakrishnan, P., Preston, E., Dueck, H., Stefanik, D.,
Tan, K., Trapnell, C., Kim, J. and Waterston, R.H., (2019).
**A lineage-resolved molecular atlas of C. elegans embryogenesis at single-cell resolution.**
*Science*, **365(6459)**. [https://doi.org/10.1126/science.aax1971](https://doi.org/10.1126/science.aax1971)

Qiu, C., Cao, J., Martin, B.K., Li, T., Welsh, I.C., Srivatsan, S., Huang, X., Calderon,
D., Noble, W.S., Disteche, C.M. and Murray, S.A. (2022). 
**Systematic reconstruction of cellular trajectories across mouse embryogenesis.** 
*Nature genetics*, **54(3)**. [https://doi.org/10.1038/s41588-022-01018-x](https://doi.org/10.1038/s41588-022-01018-x)

Sun, S., RÃ¶delsperger, C. and Sommer, R.J., (2021).
**Single worm transcriptomics identifies a developmental core network of oscillating genes with deep conservation across nematodes.**
*Genome research*, **31(9)**. [https://doi.org/10.1101/gr.275303.121](https://doi.org/10.1101/gr.275303.121)

Wolf, F.A., Angerer, P. and Theis, F.J. (2018). **SCANPY: large-scale single-cell gene expression data analysis.**
*Genome biology*, **19(1)**. [https://doi.org/10.1186/s13059-017-1382-0](https://doi.org/10.1186/s13059-017-1382-0)
