References
==========

**Bibliography**

Emms, D.M. and Kelly, S. (2019). **OrthoFinder: phylogenetic orthology inference for comparative genomics.** 
*Genome biology*, **20(1)**. https://doi.org/10.1186/s13059-019-1832-y.

Huerta-Cepas, J., Serra, F. and Bork, P. (2016). **ETE 3: reconstruction, analysis, and visualization of phylogenomic data.**
*Molecular biology and evolution*, **33(6)**. https://doi.org/10.1093/molbev/msw046.

Wolf, F.A., Angerer, P. and Theis, F.J. (2018). **SCANPY: large-scale single-cell gene expression data analysis.** 
*Genome biology*, **19(1)**. https://doi.org/10.1186/s13059-017-1382-0.

Qiu, C., Cao, J., Martin, B.K., Li, T., Welsh, I.C., Srivatsan, S., Huang, X., Calderon,
D., Noble, W.S., Disteche, C.M. and Murray, S.A. (2022). 
**Systematic reconstruction of cellular trajectories across mouse embryogenesis.** 
*Nature genetics*, **54(3)**. https://doi.org/10.1038/s41588-022-01018-x.

**Indices and tables**

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
